compression_verifier.go

```diff
@@ -0,0 +1,238 @@
+	"strings"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/golang/snappy"
+	"github.com/siddontang/go-mysql/schema"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// CompressionSnappy is used to identify Snappy (https://google.github.io/snappy/) compressed column data
+	CompressionSnappy = "SNAPPY"
+)
+
+type (
+	columnCompressionConfig map[string]string
+
+	// TableColumnCompressionConfig represents compression configuration for a
+	// column in a table as table -> column -> compression-type
+	// ex: books -> contents -> snappy
+	TableColumnCompressionConfig map[string]columnCompressionConfig
+)
+
+// UnsupportedCompressionError is used to identify errors resulting
+// from attempting to decompress unsupported algorithms
+type UnsupportedCompressionError struct {
+	table     string
+	column    string
+	algorithm string
+}
+
+func (e UnsupportedCompressionError) Error() string {
+	return "Compression algorithm: " + e.algorithm +
+		" not supported on table: " + e.table +
+		" for column: " + e.column
+}
+
+// CompressionVerifier provides support for verifying the payload of compressed columns that
+// may have different hashes for the same data by first decompressing the compressed
+// data before fingerprinting
+type CompressionVerifier struct {
+	logger *logrus.Entry
+
+	supportedAlgorithms     map[string]struct{}
+	tableColumnCompressions TableColumnCompressionConfig
+}
+
+// GetCompressedHashes compares the source data with the target data to ensure the integrity of the
+// data being copied.
+//
+// The GetCompressedHashes method checks if the existing table contains compressed data
+// and will apply the decompression algorithm to the applicable columns if necessary.
+// After the columns are decompressed, the hashes of the data are used to verify equality
+func (c *CompressionVerifier) GetCompressedHashes(db *sql.DB, schema, table, pkColumn string, columns []schema.TableColumn, pks []uint64) (map[uint64][]byte, error) {
+	c.logger.WithFields(logrus.Fields{
+		"tag":   "compression_verifier",
+		"table": table,
+	}).Info("decompressing table data before verification")
+
+	tableCompression := c.tableColumnCompressions[table]
+
+	// Extract the raw rows using SQL to be decompressed
+	rows, err := getRows(db, schema, table, pkColumn, columns, pks)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	// Decompress applicable columns and hash the resulting column values for comparison
+	resultSet := make(map[uint64][]byte)
+	for rows.Next() {
+		rowData, err := ScanByteRow(rows, len(columns)+1)
+		if err != nil {
+			return nil, err
+		}
+
+		pk, err := strconv.ParseUint(string(rowData[0]), 10, 64)
+		if err != nil {
+			return nil, err
+		}
+
+		// Decompress the applicable columns and then hash them together
+		// to create a fingerprint. decompressedRowData contains a map of all
+		// the non-compressed columns and associated decompressed values by the
+		// index of the column
+		decompressedRowData := [][]byte{}
```

Comment:
Seems like you already know exactly how long this array will be. Isn't it more idiomatic to `make` it here than rather than to `append` below? Seems more memory-efficient (fewer allocations). But probably no big deal and I honestly don't know which one is preferable. Just curious.

Reply:
We don't actually know the number of rows until iterating through all of them. That would be more efficient though if we did.